elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,187 @@
+ * |    Expression     |             SQL             | WHERE-clause Promotion |
+ * +-------------------+-----------------------------+------------------------+
+ * | H1 AND W1 AND H2  | WHERE W1 HAVING (H1 AND H2) | No                     |
+ * | (H1 OR H2) AND W1 | WHERE W1 HAVING (H1 OR H2)  | No                     |
+ * | H1 OR W1          | HAVING (H1 OR W1)           | Yes                    |
+ * | (W1 AND H1) OR W2 | HAVING ((W1 AND H1) OR W2)  | Yes                    |
+ * +-------------------+-----------------------------+------------------------+
+ * </pre>
+ * Note that {@link SplitFilterExpressionVisitor} might incur more-than-expected network I/O in the case of WHERE-clause
+ * promotion.
+ * <p>
+ * {@link SplitFilterExpressionVisitor} splits by storing {@code WHERE} and {@code HAVING} clauses in
+ * {@link Pair#getLeft() left} and {@link Pair#getRight() right} of a {@link Pair}, respectively. For example:
+ * <pre>
+ * {@code
+ * Pair<FilterExpression, FilterExpression> filterPair = filterExpression.accept(splitFilterExpressionVisitor);
+ *
+ * FilterExpression whereClauseFilter = filterPair.getLeft();
+ * FilterExpression havingClauseFilter = filterPair.getRight();
+ * }
+ * </pre>
+ * {@link SplitFilterExpressionVisitor} is thread-safe and can be accessed by multiple threads at the same time.
+ */
+@Slf4j
+public class SplitFilterExpressionVisitor implements FilterExpressionVisitor<WhereHaving> {
+
+    @Getter(value = AccessLevel.PRIVATE)
+    private final EntityDictionary entityDictionary;
+    @Getter(value = AccessLevel.PRIVATE)
+    private final FilterExpressionNormalizationVisitor normalizationVisitor;
+
+    /**
+     * Constructor.
+     *
+     * @param entityDictionary  Object that offers annotation information about an entity field
+     * @param normalizationVisitor  A helper {@link FilterExpressionVisitor} that is leveraged to transform
+     * expressions like {@code NOT (A AND B)} into {@code NOT A OR NOT B}
+     */
+    public SplitFilterExpressionVisitor(
+            final EntityDictionary entityDictionary,
+            final FilterExpressionNormalizationVisitor normalizationVisitor
+    ) {
+        this.entityDictionary = entityDictionary;
+        this.normalizationVisitor = normalizationVisitor;
+    }
+
+    @Override
+    public WhereHaving visitPredicate(final FilterPredicate filterPredicate) {
+        return isHavingPredicate(filterPredicate)
+                ? WhereHaving.pureHaving(filterPredicate) // this filterPredicate belongs to a HAVING clause
+                : WhereHaving.pureWhere(filterPredicate); // this filterPredicate belongs to a WHERE clause
+    }
+
+    @Override
+    public WhereHaving visitAndExpression(final AndFilterExpression expression) {
+        WhereHaving left = expression.getLeft().accept(this);
+        WhereHaving right = expression.getRight().accept(this);
+
+        if (left.isWhereExpression() && right.isWhereExpression()) {
```

Comment:
Can we just do 
```
return new WhereHaving(and(left.getWhere, right.getWhere), and(left.getHaving, right.getHaving))
```
I think the `and()` and `WhereHaving constructor` should be able to handle `null`